chalk-slg/src/context/mod.rs

```diff
@@ -264,3 +266,11 @@ pub trait UniverseMap<C: Context>: Clone + Debug {
 pub trait UnificationResult<C: Context> {
     fn into_ex_clause(self, ex_clause: &mut ExClause<C>);
 }
+
+pub trait AnswerStream<C: Context> {
+    fn peek_answer(&mut self) -> Option<SimplifiedAnswer<C>>;
+    fn next_answer(&mut self) -> Option<SimplifiedAnswer<C>>;
+
+    /// True if there are any strands that could yield an answer invalidating `subst`
+    fn may_invalidate(&mut self, subst: &C::Substitution) -> bool;
```

Comment:
In the callback variant, this would become:

```rust
/// Invokes `test` with each possible future answer, returning true immediately
/// if we find any answer for which `test` returns true.
fn any_future_answer(
    &mut self, 
    test: impl FnMut(&C::Substitution, &mut C::InferenceTable) -> bool,
) -> bool;
```

Reply:
Er I forgot to follow up on this. The main thing is that we would invoke with a closure which *still* calls your existing "invalidate" code, but that closure has access to the expected substitution from its environment (along with its inference table). It seems a bit more "neat" to thread down just a closure though.